clients/java/zts/core/src/main/java/com/yahoo/athenz/zts/ZTSClientCache.java

```diff
@@ -0,0 +1,285 @@
+/*
+ * Copyright 2016 Yahoo Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.yahoo.athenz.zts;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.ehcache.Cache;
+import org.ehcache.CacheManager;
+import org.ehcache.config.CacheConfiguration;
+import org.ehcache.config.builders.CacheConfigurationBuilder;
+import org.ehcache.config.builders.CacheManagerBuilder;
+import org.ehcache.impl.copy.ReadWriteCopier;
+import org.ehcache.spi.serialization.Serializer;
+import org.ehcache.spi.serialization.SerializerException;
+import org.ehcache.xml.XmlConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.charset.CharsetDecoder;
+import java.nio.charset.CharsetEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.AbstractMap;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Consumer;
+
+/**
+ * Provides optional caching for {@link ZTSClient}. <br>
+ * This class is designed to support multiple caches - but is currently only support a
+ *  cache from {domain+principal} to {@link RoleAccess} - see
+ *  {@link com.yahoo.athenz.zts.ZTSClient#getRoleAccess(java.lang.String, java.lang.String)}.  <br><br>
+ *
+ * TODO: Henry - what about:
+ *      {@link ZTSClient#getAccess(java.lang.String, java.lang.String, java.lang.String)}
+ *      {@link ZTSClient#getResourceAccess(java.lang.String, java.lang.String, java.lang.String, java.lang.String)} ?
+ *
+ * TODO: Henry - no unit-tests - need to discuss
```

Comment:
It should be very simple to add the unit tests for the caching case. We already provide a setZTSRDLGeneratedClient method where you can just pass a mock for the zts client which can just return the data first time it's called but throws an exception when the second time it's called. with caching we should only get it called once and we should never get an exception.